File: cpp/acronym/acronym.cpp
#include "acronym.h"

namespace acronym {

std::string clean(const std::string &input);
std::vector<std::string> split(const std::string &input);

std::string acronym(const std::string &phrase) {
  const std::vector<std::string> words = split(clean(phrase));
  std::string result;

  for (const std::string &word : words) {
    if (word.length() > 0) {
      result += std::toupper(word.at(0));
    }
  }

  return result;
}

std::string clean(const std::string &input) {
  const std::regex pattern("[^a-zA-Z\\s\\']");
  const std::string place_holder = " ";
  return std::regex_replace(input, pattern, place_holder);
}

std::vector<std::string> split(const std::string &input) {
  const int dismatch = -1;
  std::vector<std::string> words;
  std::regex pattern("\\s+");

  std::sregex_token_iterator iterator(input.begin(), input.end(), pattern,
                                      dismatch);
  std::sregex_token_iterator end;

  while (iterator != end) {
    words.push_back(*iterator);
    iterator++;
  }

  return words;
}

} // namespace acronym

File: go/acronym/acronym.go
// package acronym it is a module for creating acronyms. It provides a single
// function that is used to generate an acronym from an input text.
package acronym

import (
	"regexp"
	"strings"
)

// Abbreviate returns an acronym of the given string
func Abbreviate(s string) string {
	return strings.ToUpper(strings.Join(initialsOf(s), ""))
}

// words is a pattern of words in a sentence. Your capture group gets the
// initial of each word and the not group discards everything other than
// the initial.
var words = regexp.MustCompile(`([a-zA-Z])(?:[a-zA-Z\'])*`)

// initialsOf returns a slice containing all the initials of the words
// contained in the given string.
func initialsOf(s string) []string {
	initials := []string{}

	for _, match := range words.FindAllStringSubmatch(s, -1) {
		initials = append(initials, match[1])
	}

	return initials
}
